### DISCUSSION TITLE
增量索引合并

### DISCUSSION THREAD

**Discussion by mykittylover** [milvus-io/milvus#33871] [Q&A and General discussion]
索引构建分全量索引和增量索引，全量构建使用的是全量数据构建，增量构建使用的是不分数据构建，当增量构建后，如何与全量索引合并一下，然后加载新的增量索引版本，以此来减少索引的内存占用，提升检索效率

**Comment by yhmo** [comment]
**请下次提问时尽量使用英文。**

Milvus是按分片(segment)管理数据，每个分片都有独立的索引，并不是全量数据构建索引。

**Reply by mykittylover** [reply]
"Alright, got it. Then, how are segments merged? Because having too many segments will inevitably lead to a decrease in search efficiency and also cause memory to continuously increase."

**Reply by yhmo** [reply]
There is an internal compaction machinery, detects and merges small segments into large segments.

Let's say there are 2 small segments A and B. Both of A and B have been indexed.
The compaction workflow:
1. detects 2 small segments A and B
2. data node loads A and B into memory
3. data node merges A and B in memory, generates a large segment C and persist to storage
4. index node detects there is a new segment C
5. index node loads C into memory, builds index for C, the index of C is persist to storage
6. once C is indexed, the A and B are marked as "to-be-delete"
7. GC process triggers, A and B are deleted, along with their indexes
8. If query node already loads A and B in memory, the query node will load C index into memory to replace the indexes of A and B

**Reply by yhmo** [reply]
With small segments are compacted into large segments, and large segments are indexed. Eventually, large index replace the small indexes and search performance increase, memory usage no change since small indexes will be evicted from memory.

**Reply by mykittylover** [reply]
"I am referring to index merging as the process of combining already constructed index data that is segmented into multiple parts. These segments may contain duplicate data, and I need to merge them into a larger index while removing any duplicates and deleting the data that is marked for deletion."

**Reply by yhmo** [reply]
This is job of Milvus compaction machinery, already done.

**Reply by xiaofan-luan** [reply]
Milvus don't handle duplicated data. you will need to handle it in advance, or use upsert to delete old data.

compaction merged all the data, only by pk and it's ts order
